### PR TITLE
GIT-119: handle scoping for parallel tests

### LINKED ISSUE #119: TestInParallel feature not working as expected
Although the features are getting executed in parallel, but the assess written within those features are getting mixed up with other features in parallel execution.

e2e-framework version used: **v0.0.6**

**Ex:**
Feature1 -> assess1
Feature2 -> asses2
When we try to run them parallel, it executes the following
	feature1 -> assess2
	feature2 -> assess2

**Sample code:**

//go:build all || dummy

package install_tests

import (
    "context"
    "fmt"
    "os"
    "testing"

    "sigs.k8s.io/e2e-framework/pkg/env"
    "sigs.k8s.io/e2e-framework/pkg/envconf"
    "sigs.k8s.io/e2e-framework/pkg/features"
    "time"
)

var (
    clusterName string
    namespace   string
    testEnv     env.Environment
)

func TestMain(m *testing.M) {

    cfg, _ := envconf.NewFromFlags()
    testEnv = env.NewWithConfig(cfg)

    os.Exit(testEnv.Run(m))
}

func TestParallelRuns(t *testing.T) {
    featureOne := features.New("Feature One").
        Assess("Assess 1", func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
            println(fmt.Sprintf("feature1-Assess1, datetime: %s", time.Now().String()))
            time.Sleep(10 * time.Second)
            return ctx

        }).Feature()

    featureTwo := features.New("Feature Two").
        Assess("Assess 2", func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
            time.Sleep(2 * time.Second)
            println(fmt.Sprintf("feature2-Assess2, datetime: %s", time.Now().String()))
            return ctx
        }).Feature()

    testEnv.TestInParallel(t, featureOne, featureTwo)

}

**Execution cmd:** go test . -v -timeout 240m -count=1 -tags=dummy -args --parallel

**Execution output:**

=== RUN   TestParallelRuns
=== RUN   TestParallelRuns/Feature_Two
=== RUN   TestParallelRuns/Feature_Two/Assess_2
=== RUN   TestParallelRuns/Feature_One
=== RUN   TestParallelRuns/**Feature_One/Assess_2**
**feature2-Assess2**, datetime: 2022-03-16 18:06:37.489844 +0530 IST m=+2.009548369
**feature2-Assess2**, datetime: 2022-03-16 18:06:37.48979 +0530 IST m=+2.009494100
--- PASS: TestParallelRuns (2.00s)
    --- PASS: TestParallelRuns/Feature_Two (2.00s)
        --- PASS: TestParallelRuns/Feature_Two/Assess_2 (2.00s)
    --- PASS: TestParallelRuns/Feature_One (2.00s)
        --- PASS: TestParallelRuns/Feature_One/Assess_2 (2.00s)
PASS
ok      gitlab.eng.vmware.com/tap/tap-packages/suite/dummy      2.479s

**NOTE**: Notice that the Assess_2 is not part of Feature_One but is executed under it.

### LINKED ISSUE #122: Flaky Test Setup due to `kind` Setup
Seems like we are running into a few issue with the Patchset jobs due to the way we are installing kind with `go1.17`

```bash
main.main()
	_testmain.go:91 +0x28f

goroutine 20 [chan receive]:
k8s.io/klog/v2.(*loggingT).flushDaemon(0x0?)
	/home/prow/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:1181 +0x6a
created by k8s.io/klog/v2.init.0
	/home/prow/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:420 +0xf6

goroutine 21 [sleep]:
time.Sleep(0x6fc23ac00)
	/usr/local/go/src/runtime/time.go:194 +0x12e
sigs.k8s.io/controller-runtime/pkg/log.init.0.func1()
	/home/prow/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/log/log.go:63 +0x38
created by sigs.k8s.io/controller-runtime/pkg/log.init.0
	/home/prow/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/log/log.go:62 +0x25
FAIL	sigs.k8s.io/e2e-framework/examples/decoder	0.350s
F0325 19:23:43.575830    5131 env.go:333] kind not available even after installation
goroutine 1 [running]:
```

This blocks the PRs from getting merged. 

xref: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_e2e-framework/120/pull-e2e-framework-test/1507437780685295616/build-log.txt

https://kind.sigs.k8s.io/docs/user/quick-start/#installing-with-go-get--go-install suggests a minor change in installing `kind` on the cluster using `go`.

### PR DESCRIPTION
**Pull request by harshanarayana**
Closes #119 

```bash
❯ go test -v . -args --parallel
I0316 19:26:00.910966   58073 parallel_features_test.go:46] "Args" flag=&{client:<nil> kubeconfig: namespace: assessmentRegex:<nil> featureRegex:<nil> labels:map[] skipFeatureRegex:<nil> skipLabels:map[] skipAssessmentRegex:<nil> parallelTests:true}
=== RUN   TestPodBringUp
=== RUN   TestPodBringUp/Feature_Two
=== RUN   TestPodBringUp/Feature_Two/Create_Nginx_Deployment_2
=== RUN   TestPodBringUp/Feature_One
=== RUN   TestPodBringUp/Feature_One/Create_Nginx_Deployment_1
=== RUN   TestPodBringUp/Feature_Two/Wait_for_Nginx_Deployment_2_to_be_scaled_up
=== RUN   TestPodBringUp/Feature_One/Wait_for_Nginx_Deployment_1_to_be_scaled_up
--- PASS: TestPodBringUp (65.08s)
    --- PASS: TestPodBringUp/Feature_Two (65.05s)
        --- PASS: TestPodBringUp/Feature_Two/Create_Nginx_Deployment_2 (0.04s)
        --- PASS: TestPodBringUp/Feature_Two/Wait_for_Nginx_Deployment_2_to_be_scaled_up (65.01s)
    --- PASS: TestPodBringUp/Feature_One (65.07s)
        --- PASS: TestPodBringUp/Feature_One/Create_Nginx_Deployment_1 (0.07s)
        --- PASS: TestPodBringUp/Feature_One/Wait_for_Nginx_Deployment_1_to_be_scaled_up (65.01s)
PASS
ok  	sigs.k8s.io/e2e-framework/examples/parallel_features	331.487s
```

Closes #122 as well

### CODE REVIEW

**Comment by harshanarayana** [comment]
@vladimirvivien  @ShwethaKumbla PTAL

**Review by ShwethaKumbla — APPROVED** [review]
/lgtm

File: pkg/env/env.go [hunk 1]
```diff
@@ -243,6 +243,7 @@ func (e *testEnv) processTests(t *testing.T, enableParallelRun bool, testFeature
 
 	var wg sync.WaitGroup
 	for i, feature := range testFeatures {
+		feature := feature
```
Comment (vladimirvivien) [inline pkg/env/env.go:246]: I would suggest couple of things:
1. For clarity, name the new variable something else other than `feature`
2. Have your anonymous function (line 253) take a feature, as well, as parameter

```go
if runInParallel {
			wg.Add(1)
			go func(w *sync.WaitGroup, f Feature) {
				defer w.Done()
				e.processTestFeature(t, f.Name, f)
			}(&wg, feature)
		} else {
			e.processTestFeature(t, featName, feature)
		}
```

**Review by vladimirvivien — COMMENTED** [review]
Thanks for the quick fix. I left comments.

File: pkg/env/env.go (hunk 1 above)
Comment (harshanarayana) [reply pkg/env/env.go]: Thanks for the review @vladimirvivien . Addressed the requests.

**Comment by harshanarayana** [comment]
Also upticked and fixed the way the `kind` binary was getting setup to avoid test failures. 

```bash

goroutine 20 [chan receive]:
k8s.io/klog/v2.(*loggingT).flushDaemon(0x0?)
	/home/prow/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:1181 +0x6a
created by k8s.io/klog/v2.init.0
	/home/prow/go/pkg/mod/k8s.io/klog/v2@v2.30.0/klog.go:420 +0xf6

goroutine 21 [sleep]:
time.Sleep(0x6fc23ac00)
	/usr/local/go/src/runtime/time.go:194 +0x12e
sigs.k8s.io/controller-runtime/pkg/log.init.0.func1()
	/home/prow/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/log/log.go:63 +0x38
created by sigs.k8s.io/controller-runtime/pkg/log.init.0
	/home/prow/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.0/pkg/log/log.go:62 +0x25
FAIL	sigs.k8s.io/e2e-framework/examples/decoder	0.350s
F0325 19:23:43.575830    5131 env.go:333] kind not available even after installation
```

**Review by ShwethaKumbla — APPROVED** [review]
/lgtm

**Review by vladimirvivien — APPROVED** [review]
LGTM

**Comment by vladimirvivien** [comment]
@harshanarayana 
Apologies about the delay. Super busy at work.
Thanks for the help.

**Pull request merged**